tiledb/sm/subarray/range_subset.h

```diff
@@ -0,0 +1,328 @@
+        });
+    return Status::Ok();
+  };
+};
+
+template <Datatype D>
+struct SortStrategy<D, char, char> {
+  static Status sort(ThreadPool* const, std::vector<Range>&) {
+    return LOG_STATUS(Status_SubarrayError(
+        "Invalid datatype " + datatype_str(D) + " for sorting."));
+  };
+};
+
+template <>
+struct SortStrategy<Datatype::STRING_ASCII, std::string, std::string> {
+  static Status sort(ThreadPool* const compute_tp, std::vector<Range>& ranges) {
+    parallel_sort(
+        compute_tp,
+        ranges.begin(),
+        ranges.end(),
+        [&](const Range& a, const Range& b) {
+          return a.start_str() < b.start_str() ||
+                 (a.start_str() == b.start_str() && a.end_str() < b.end_str());
+        });
+    return Status::Ok();
+  };
+};
+
+}  // namespace detail
+
+class RangeSubsetBase {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /** Destructor. */
+  virtual ~RangeSubsetBase() = default;
+
+  // /**
+  //  * Adds a range to the range manager. If a default range manager strategy
+  //  is,
+  //  * then first update the range strategy.
+  //  *
+  //  * @param ranges The current ranges in the subarray (remove after
+  //  refactor).
+  //  * @param new_range The range to add.
+  //  */
+  // // virtual Status add_range(
+  // //   std::vector<std::vector<Range>>& ranges, Range&& new_range) = 0;
+
+  /**
+   * Adds a range to the range manager without performing any checkes. If a
+   * default strategy is set, then first update the range strategy.
+   *
+   * @param ranges The current ranges in the subarray (remove after
+   refactor).
+   * @param new_range The range to add.
+   */
+  virtual Status add_range_unsafe(const Range& range) = 0;
+
+  virtual const Range& get_range(const uint64_t range_index) const = 0;
+
+  virtual const std::vector<Range>& get_ranges() const = 0;
+
+  /**
+   * Returns ``true`` if the current range is the default range.
+   *
+   * The default range is the full range for the dimension the range is
+   * managing.
+   **/
+  virtual bool is_default() const = 0;
+
+  /** Returns ``true`` if the range subset is the empty set. */
+  virtual bool is_empty() const = 0;
+
+  /**
+   * Returns ``false`` if the subset contains a range other than the default
+   * range.
+   **/
+  virtual bool is_set() const = 0;
+
+  /**
+   * Returns ``true`` if there is exactly one unary ranage in the subset.
+   */
+  virtual bool is_unary() const = 0;
+
+  /** Returns the number of distinct ranges stored in the range manager. */
+  virtual uint64_t num_ranges() const = 0;
+
+  /**
+   * Sorts the ranges in the range manager.
+   *
+   * @param compute_tp
+   */
+  virtual Status sort_ranges(ThreadPool* const compute_tp) = 0;
+};
+
+template <typename T, Datatype D, bool CoalesceAdds>
+class RangeSubset : public RangeSubsetBase {
+ private:
+  using AddStrategy = detail::AddStrategy<CoalesceAdds, T>;
```

Comment:
The heart of variation in the subclasses is in this declaration. If a member function here references it, it's varying between subclasses. If it doesn't, then it can go in the base class. Ditto for `SortStrategy`.